Major Refactor + New Features

AbhijeetKrishnan opened the pull request:
_(not expecting this to get merged as is, just starting the conversation)_

This is a major refactor of the codebase to introduce more logical separation of components, more testability, and overall a better foundation for future development.

The redesign separates the major components of the bot into the FrameDataBot (bot-specific code, command tree, generating embeds), FrameDb (in-memory frame data db, querying) and FrameService (fetching frame data from external API, cleaning data). It also adds tests for each of these, although many are unimplemented. It adds a stricter dev workflow using `ruff` for linting + formatting and `mypy` for type checking.

Also included are some extra, possibly unwanted features -

- Autocomplete for character names
- Searching for a move by move name
- Character portraits based on Wavu wiki thumbnails
- Removal of `/character` commands, unifying everything under just `/fd`
- Addition of `/help` command

<s>Lastly, the current build doesn't pass because of a failing test related to an Asuka move for which I'm not sure what the expected behavior is.</s>

TLNBS2405 commented:
Can you also update the readme? It seems there are some additional libraries there, right?

AbhijeetKrishnan commented:
> Can you also update the readme? It seems there are some additional libraries there, right?

Yes, I've added [`fast-autocomplete`](https://pypi.org/project/fast-autocomplete/) for the auto-completion feature when searching for characters. I don't see where the requirements are mentioned in the original README though. At most, the instructions to setup the bot are mentioned with `pip install -r requirements.txt`. I've updated the installation instructions for the refactor. `pip` automatically picks up the requirements from the `pyproject.toml` file.
